Fix #18648: Include Newsletter and update the Footer Section

shricodev opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #18648.
2. This PR does the following: Add  a newsletter section and tweak footer design. Remove the Join Us form in the donate page and move the thanks for subscribing modal inside base-components. Also fix the case when some users were having name "null" stored in the backend.
3. Link to the Figma file: https://www.figma.com/file/O3mJt26w9CGOnSKlUS4RXy/Oppia-Footer-and-Newsletter?type=design&node-id=275-516&mode=design&t=4IKln54vO3x8bRKJ-0

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface.

Please also include videos/screenshots of the developer tools
browser console, so that we can be sure that there are no console errors.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

![image](https://github.com/oppia/oppia/assets/76906722/b10fc7e3-d1f5-4b29-b3c1-3ff1e832c08e)



https://github.com/oppia/oppia/assets/76906722/8c43750d-a287-4a74-9eaa-93797baa88f6




Remove the Join Us form from the donate page as this would be redundant.

![Screenshot 2023-07-24 091210](https://github.com/oppia/oppia/assets/76906722/2ff21568-6b1c-4226-9c8c-c58c277396ba)

#### Proof of changes on mobile phone


https://github.com/oppia/oppia/assets/76906722/b6180d6d-2eaa-4e3d-a5c1-107d6612b23d



## Proof of Changes in Arabic Language


![image](https://github.com/oppia/oppia/assets/76906722/4b62d93f-0af1-4b17-9e71-f93f80ac41f5)



<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- Make sure your PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

shricodev commented:
Message for Reviewers: I am new to angular so might have made some silly mistakes. Kindly, ask for changes if needed. Thanks!

shricodev commented:
@lkbhitesh07 @Lawful2002 @srijanreddy98 @seanlip PTAL

seanlip commented:
Thanks @shricodev! Two things:
- Can you show a video that clicks through the different pages on Oppia and shows the footer? Include some pages which don't have a footer, too.
- I think it would be good for @ConorMack to review the screenshots/videos since he designed the updated footer and will have better comments on the UX, so I'm assigning him as a reviewer as well!

shricodev commented:
> Thanks @shricodev! Two things:
> 
> * Can you show a video that clicks through the different pages on Oppia and shows the footer? Include some pages which don't have a footer, too.
> * I think it would be good for @ConorMack to review the screenshots/videos since he designed the updated footer and will have better comments on the UX, so I'm assigning him as a reviewer as well!

Sure @seanlip, Done

Lawful2002 commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+        );
+      } else {
+        this.alertsService.addInfoMessage(
+          'Sorry, an unexpected error occurred. Please email admin@oppia.org ' +
+          'to be added to the mailing list.', 10000);
```
Would it be better to add this error message to the constants file, and reference it here and in the unit tests?

shricodev replied:
Done! I will push the changes here after I address Sean's review :). Thanks

Lawful2002 commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+      }
+    }).catch(errorResponse => {
+      this.alertsService.addInfoMessage(
+        'Sorry, an unexpected error occurred. Please email admin@oppia.org ' +
+        'to be added to the mailing list.', 10000);
```
Ditto

shricodev replied:
Done, Thanks

Lawful2002 commented on core/templates/pages/donate-page/donate-page.component.spec.ts:
```diff
@@ -69,7 +68,7 @@ describe('Donate page', () => {
   let pageTitleService: PageTitleService;
   let windowRef: MockWindowRef;
-  let mailingListBackendApiService: MailingListBackendApiService;
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
Why is this needed here?

Lawful2002 commented:
Please reassign when this PR is ready for review.

lkbhitesh07 commented:
Please assign me again after the PR is ready for review.
Thanks

U8NWXD commented:
Replacing @srijanreddy98 with @seanlip and @heyimShivam since @srijanreddy98 is away

seanlip commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+  subscribeToMailingList(): void {
+    this.mailingListBackendApiService.subscribeUserToMailingList(
+      String(this.emailAddress),
+      // Most of the times, since the name is not provided
```
Most of the time, the name is `null` since it is not provided.

shricodev replied:
Thanks, Done

seanlip commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+      String(this.emailAddress),
+      // Most of the times, since the name is not provided
+      // so it is null. In such case placeholder name is set to UserX.
+      String(this.name) ?? 'UserX',
```
This seems bad. Don't pass fake data to the API. Just send a null name.

shricodev replied:
I was also not sure about this approach, thanks for pointing it out. :)

seanlip commented on core/templates/base-components/thanks-for-subscribing-modal.component.ts:
```diff
@@ -13,13 +13,12 @@
 
 /**
- * @fileoverview Controller for the donation page thanks for subscribing modal.
+ * @fileoverview Controller for the thanks for subscribing modal.
```
Maybe add hyphens: thanks-for-subscribing

shricodev replied:
Thanks, Done

seanlip commented on core/templates/base-components/oppia-footer.component.css:
```diff
@@ -6,13 +6,22 @@
   margin-left: auto;
   margin-right: auto;
-  max-width: 800px;
+  max-width: 1150px;
```
Just to check, what's the motivation behind this specific value?

shricodev replied:
I had to increase the value to fit the Oppia mission text. I think this amount works best because it helps us make the most of the space and keeps things nice and tidy.

seanlip replied:
Yup, but how was 1150 determined (as opposed to 1120 or 1180, say)?

shricodev replied:
I am not sure if I will be able to answer it. This looked a bit more similar to the Figma file provided and also it seemed to maintain a similar spacing.

seanlip commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+      AppConstants.MAILING_LIST_WEB_TAG
+    ).then((status) => {
+      if (status) {
+        this.alertsService.addInfoMessage('Done!', 1000);
```
I'm not entirely sure if this will lead to the right UI. Could you please show videos with each of these info messages showing so we can see the user experience?

shricodev replied:
This seems to work properly with proper UI. I am attaching a video of it with different info messages. Please have a look

https://github.com/oppia/oppia/assets/76906722/878e0ed6-d9eb-4d75-85e1-191b4c4124b5

PS: Not a part of this PR though, but the `thanks-for-subscribing-modal` does not seem right in the dimension `768-991`.

seanlip replied:
This looks good, thanks. Could you file a separate issue for the modal problem please?

seanlip replied:
This looks good, thanks. Could you file a separate issue for the modal problem please?

shricodev replied:
Done, opened an issue for it. also, the same issue seemed to happen in the `donating-modal`. #18709

seanlip commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -53,6 +63,40 @@ export class OppiaFooterComponent {
+    return regex.test(String(this.emailAddress));
+  }
+
+  subscribeToMailingList(): void {
```
I have some concerns about this:

- We removed this functionality from the donate page. So why isn't it deleted in this PR?
- How did you test that this works? There should be frontend tests, but I don't see a spec file.

shricodev replied:
I am sorry :(, but I am not able to quite understand which functionality you are talking about `subscribeToMailingList` or `validateEmailAddress`. Could you please elaborate a bit more and the test file?

seanlip replied:
Yup. You removed the newsletter section from the HTML in the donate page. So shouldn't there be some removal of newsletter-signup TypeScript code (that should now be dead code since it's no longer used)?

shricodev replied:
Thank You. Sorry, I am not able to find any unused code in `donate-page.component.ts` and `donate-page.component.spec.ts`. Could you please point me to any unused code? I believe I removed the sign-up ts logic from both files.

seanlip reviewed (changes requested):
Thanks @shricodev! Left some comments. I wasn't sure about some parts of the UI (namely the email part seems a bit wide) but I'm discussing them with @ConorMack.

seanlip commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -72,9 +72,8 @@ export class OppiaFooterComponent {
-      // so it is null. In such case placeholder name is set to UserX.
-      String(this.name) ?? 'UserX',
+      // Most of the time, the name is null since it is not provided.
+      String(this.name),
```
I did a bit of digging and what I'm finding worries me ... it seems our "add_user_to_mailing_list" logic is a bit off and we have some users in the mailing list with name "null" because of that. Here is what should happen instead.

- If a name is not supplied, a null/None value for the name field should be passed to the backend instead.
- In the backend, this null value should be carried through, and not replaced with a string at any point.
- When we get to "add_user_to_mailing_list()" in user_services.py, the name field should be optional and default to None (so, move the "name" argument after the "tag" argument and fix all callsites). Any other calls to "add_user_to_mailing_list()" that supply an explicit name should have `name=` added before that argument since it now has a default value. Any calls that don't supply an explicit name should not pass that argument.
- In "add_user_to_mailing_list()", if the name is None then the merge_fields for name should not be provided to bulk_email_services.

Could you PTAL at this pipeline? We should fix this once and for all, otherwise this confusion about null/non-null names is going to continue.

shricodev replied:
I will look into it and maybe update the PR by tomorrow evening.

shricodev replied:
I have made some necessary changes, PTAL

lkbhitesh07 commented on assets/i18n/en.json:
```diff
@@ -616,13 +616,20 @@
     "I18N_FOOTER_BROWSE_LIBRARY": "Browse the Library",
     "I18N_FOOTER_CONTACT_US": "Contact Us",
     "I18N_FOOTER_CONTRIBUTE_ALL_CAPS": "CONTRIBUTE",
+    "I18N_FOOTER_COPYRIGHT": "© 2023 The Oppia Foundation",
```
I think we should not hardcore the year, we can maybe add an angular logic to get the current year. What do you think?

shricodev replied:
Indeed, that would help in code maintainability. I have modified it. PTAL

lkbhitesh07 reviewed (changes requested):
Thanks for the changes @shricodev, Just one inline comment for the codeowners files.

heyimShivam commented on core/templates/pages/donate-page/donate-page.component.spec.ts:
```diff
@@ -69,7 +68,7 @@ describe('Donate page', () => {
   let pageTitleService: PageTitleService;
   let windowRef: MockWindowRef;
-  let mailingListBackendApiService: MailingListBackendApiService;
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
```
These changes are really required.?

shricodev replied:
Oh, the `alertsService` is currently not in use and therefore I am removing it from the test file. Thanks

heyimShivam reviewed (approved):
LGTM

Lawful2002 reviewed (approved):
Thanks @shricodev, the frontend tests LGTM!
EDIT: A few comments have not been answered.

Lawful2002 reviewed (commented):
A few un-answered comments.

lkbhitesh07 reviewed (approved):
Thanks for the changes @shricodev, LGTM for the code owner files.

seanlip commented on core/controllers/profile.py:
```diff
@@ -253,7 +253,7 @@ def put(self) -> None:
         name = self.normalized_payload['name']
         tag = self.normalized_payload['tag']
-        status = user_services.add_user_to_mailing_list(email, name, tag)
+        status = user_services.add_user_to_mailing_list(email, tag, name)
```
This should be add_uiser_to_mailing_list(email, tag, name=name).

(For default args, always include the arg name at the callsite. Please fix anywhere else that's appropriate.)

shricodev replied:
Hey, I'm having a bit of a problem. After changing this with the keyword argument `name=name` I had to also change the test file for the `test_put_function` but don't know why I am not able to correctly update the test file and ended up failing `backend_tests`. :(

So, I decided to not push those changes to the test file. Any help would be highly appreciated. Thanks

seanlip replied:
I don't know either because I don't have enough information. I suggest that you create a [debugging doc](https://docs.google.com/document/d/1qRbvKjJ0A7NPVK8g6XJNISMx_6BuepoCL7F2eIfrGqM/edit) and share it, with information about what you tried and why.

shricodev replied:
I have been trying to fix this issue for a few days but since it is a keyword argument. the `swap` function had to be changed to also accept the keyword argument, I modified it as well. Don't know if because of this change the app is breaking somehow and the backend test is failing.


https://github.com/oppia/oppia/blob/622e833ee29cbb35e4e3e28b484503fc7b6f0ca4/core/controllers/profile_test.py#L988-L989

I was wondering, do we need to handle it this seriously for such non non-breaking change? I know this sounds bad but.. :(

shricodev replied:
I apologize for the late response to the review. :( I actually made a silly mistake of not including `**kwargs` in the test code `test_put_function`. Now, everything should work as expected and the test should pass.  PTAL, Thanks

seanlip commented on core/domain/user_services.py:
```diff
@@ -1744,21 +1745,23 @@ def record_user_created_an_exploration(user_id: str) -> None:
         email: str. Email of the user.
-        name: str. Name of the user.
         tag: str. Tag for the mailing list.
+        name: None or str. Name of the user.
```
str or None. Name of the user, or None if no name was supplied.

shricodev replied:
Done, Thanks

seanlip commented on core/platform/bulk_email/mailchimp_bulk_email_services.py:
```diff
@@ -253,14 +253,14 @@ def add_or_update_user_status(
             'tags': [tag],
             'merge_fields': {
-                'NAME': merge_fields['NAME']
+                'NAME': merge_fields.get('NAME', None)
```
I think, if no name is provided, then just don't send this field? Ditto below.

shricodev replied:
Done, Thanks

seanlip reviewed (changes requested):
OK, I think this looks good just minor comments left. Thanks @shricodev!

Also @rfinch11, would you mind doing a UI pass please since @ConorMack isn't around atm? Thanks!

rfinch11 commented:
> OK, I think this looks good just minor comments left. Thanks @shricodev!
> 
> Also @rfinch11, would you mind doing a UI pass please since @ConorMack isn't around atm? Thanks!

This is the first time I'm seeing this button color on the site. Are we using the orange button color elsewhere? I suggest switching to the secondary button style as specified in the [design system](https://www.figma.com/file/nrQrOxSxSg1oabuVE6XJpa/Oppia-Design-Guide?type=design&node-id=0%3A1&mode=design&t=5iDEijkWYvvTitR0-1
) 
<img width="816" alt="Screen Shot 2023-08-03 at 11 14 47 AM" src="https://github.com/oppia/oppia/assets/17371709/44495512-2c3f-45ea-b4e0-8312dc41aea6">

I also suggest changing the button text to "Subscribe" to match the language in the header. "Join now" could have multiple interpretations and be less clear to the user. Maybe @roxanashirazi could take a look from UXW perspective?

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -1,6 +1,37 @@
+        <p [innerHTML]="'I18N_FOOTER_NEWSLETTER_TEXT' | translate"></p>
+      </div>
+      <div class="oppia-footer-newsletter-input">
+        <label id="email"
+               [innerHTML]="'I18N_FOOTER_NEWSLETTER_EMAIL' | translate"
+        >
+        </label>
```
```suggestion
        <label id="email">
            {{ 'I18N_FOOTER_NEWSLETTER_EMAIL' | translate }}
        </label>
```
Prefer to use interpolation, here and elsewhere.

shricodev replied:
Done, Thanks

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -1,6 +1,37 @@
+          <button class="oppia-footer-newsletter-button"
+                  [disabled]="!validateEmailAddress()"
+                  (click)="subscribeToMailingList()">
+            <span [innerHTML]="'I18N_FOOTER_NEWSLETTER_SUBSCRIBE' | translate"></span>
```
Ditto

shricodev replied:
Done, Thanks

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -1,6 +1,37 @@
   <div class="oppia-footer-container">
     <div class="row row-cols-2">
+      <div class="oppia-mission-volunteer col-5">
+        <p class="oppia-mission" [innerHTML]="'I18N_FOOTER_OPPIA_MISSION' | translate"></p>
+        <p class="oppia-donate-volunteer" [innerHTML]="'I18N_FOOTER_OPPIA_MISSION_DONATE_VOLUNTEER' | translate"></p>
```
Ditto

Lawful2002 replied:
This?

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -1,6 +1,37 @@
+      </div>
+      <hr class="oppia-footer-separator">
       <div class="col">
         <p [innerHTML]="'I18N_FOOTER_ABOUT_ALL_CAPS' | translate"></p>
```
Ditto

shricodev replied:
Hey, When using interpolation in `oppia-donate-volunteer`, the tags are being interpreted as is. As a result, I had to stick with it. Thanks

Lawful2002 replied:
Does this not work? `<p class="oppia-donate-volunteer">{ 'I18N_FOOTER_OPPIA_MISSION_DONATE_VOLUNTEER' | translate }</p>`

shricodev replied:
I am sorry if I couldn't explain it properly to you :(. I tried doing what you suggested earlier. but on doing so. the tags in the `i18n` values were rendered as is.

![Screenshot 2023-08-06 090425](https://github.com/oppia/oppia/assets/76906722/1170b322-5b96-422d-b45f-00c152ad685d)

But with the `[innerHtml]` approach they seem to render as HTML. Correct me if I am wrong, Thanks

Lawful2002 replied:
Umm... Weird can you share the piece of code that causes this issue with internpolation? (also looks like I made a mistake in the comment, use a double {{ }} to enclose the data to be interpolated)

shricodev replied:
```html
      <div class="oppia-mission-volunteer col-7">
        <p class="oppia-mission">{{ 'I18N_FOOTER_OPPIA_MISSION' | translate }}</p>
        <p class="oppia-donate-volunteer">{{ 'I18N_FOOTER_OPPIA_MISSION_DONATE_VOLUNTEER' | translate }}</p>
      </div>
```

In this code, regardless of whether we do it, the i18n will be rendered as plain text. So, in this scenario, any tags in the i18n

```json
"I18N_FOOTER_OPPIA_MISSION_DONATE_VOLUNTEER": "<a href=\"/donate\">Donate</a> or <a href=\"/volunteer\">Volunteer</a> today!",
```

They will, however, be shown as HTML if you use `[innerHtml]`. So, I believe we should use `[innerHtml]` in this situation. What do you think?

Lawful2002 replied:
Ah, I see, yes that is correct, didn't know that the variable pointed to HTML code.

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -112,4 +105,21 @@
     </div>
   </div>
+  <div class="oppia-footer-terms-privacy">
+    <small [innerHTML]="'I18N_FOOTER_COPYRIGHT' | translate:{currentYear: currentYear}"></small>
```
Ditto

shricodev replied:
Done, Thanks

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -112,4 +105,21 @@
+      <a href="/terms"
+         [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.TERMS.ROUTE"
+         class="e2e-test-terms-link"
+         [innerHTML]="'I18N_FOOTER_TERMS_OF_SERVICE' | translate">
```
Ditto

shricodev replied:
Done, Thanks

Lawful2002 commented on core/templates/base-components/oppia-footer.component.html:
```diff
@@ -112,4 +105,21 @@
+      <a href="/privacy-policy"
+         [smartRouterLink]="'/' + PAGES_REGISTERED_WITH_FRONTEND.PRIVACY.ROUTE"
+         class="e2e-test-privacy-policy-link"
+         [innerHTML]="'I18N_FOOTER_PRIVACY_POLICY' | translate">
```
Ditto

shricodev replied:
Done, Thanks

Lawful2002 reviewed (changes requested):
Thanks, left a few comments.

Lawful2002 commented:
Replied to comments, PTAL. Thanks!

seanlip reviewed (changes requested):
Hi @shricodev, please respond to all comments before asking for a review. I think this one is left: https://github.com/oppia/oppia/pull/18694#discussion_r1283528783 -- and also  the top-level comment by @rfinch11: https://github.com/oppia/oppia/pull/18694#issuecomment-1664431676

Thanks!

shricodev commented:
> > OK, I think this looks good just minor comments left. Thanks @shricodev!
> > Also @rfinch11, would you mind doing a UI pass please since @ConorMack isn't around atm? Thanks!
> 
> This is the first time I'm seeing this button color on the site. Are we using the orange button color elsewhere? I suggest switching to the secondary button style as specified in the [design system](https://www.figma.com/file/nrQrOxSxSg1oabuVE6XJpa/Oppia-Design-Guide?type=design&node-id=0%3A1&mode=design&t=5iDEijkWYvvTitR0-1) <img alt="Screen Shot 2023-08-03 at 11 14 47 AM" width="816" src="https://user-images.githubusercontent.com/17371709/258200085-44495512-2c3f-45ea-b4e0-8312dc41aea6.png">
> 
> I also suggest changing the button text to "Subscribe" to match the language in the header. "Join now" could have multiple interpretations and be less clear to the user. Maybe @roxanashirazi could take a look from UXW perspective?

Hello @rfinch11, I have implemented the new Figma design and incorporated your suggestion for the button color. This is how it looks now.


https://github.com/oppia/oppia/assets/76906722/51d6b2e6-9693-4b2a-8297-ec4f9a14cb59

Due to the green background, the button boundaries are not very visible. I would appreciate your insights on this. Thank you.

rfinch11 commented:
>  Recording.2023-08-06.214611.mp4 
> Due to the green background, the button boundaries are not very visible. I would appreciate your insights on this. Thank you.

@shricodev that's a good point about the boundaries not being as noticeable. The other option is to implement the same style as the language picker in the header, since the background color is the same. 

<img width="153" alt="Screen Shot 2023-08-08 at 9 35 11 AM" src="https://github.com/oppia/oppia/assets/17371709/5eb61f02-35b4-4d4f-88cf-728aee6724c7">

shricodev commented:
> > Recording.2023-08-06.214611.mp4
> > Due to the green background, the button boundaries are not very visible. I would appreciate your insights on this. Thank you.
> 
> @shricodev that's a good point about the boundaries not being as noticeable. The other option is to implement the same style as the language picker in the header, since the background color is the same.
> 
> <img alt="Screen Shot 2023-08-08 at 9 35 11 AM" width="153" src="https://user-images.githubusercontent.com/17371709/259168068-5eb61f02-35b4-4d4f-88cf-728aee6724c7.png">

@rfinch11  Yeah, I have now changed it according to the dropdown, but I made it active instead of changing the BG color on hover.



https://github.com/oppia/oppia/assets/76906722/4a756d2d-278a-4848-b26d-9c83079c6e6f

ConorMack commented:
Hey All! 

I apologize for the late reply on this. I have had some family issues and took the past few weeks off. **It looks like you've gotten a lot done, great job!** 

The video of the breakpoints looks great. Also, @rfinch11 suggestion to use the button styling from the dropdown was a great idea. I think it looks better than the original design. 

I did update the design file to tweak alignments. The newsletter signup, footer content and page content above it should all be the same width. Check it out and ask if you have any questions. Just make sure to tag me in any comments you need a direct response from. 

**-Screenshot-**
<img width="612" alt="Screenshot 2023-08-09 at 10 39 56 AM" src="https://github.com/oppia/oppia/assets/11729306/b141acdf-58d3-4774-9b78-f106b6126ad9">


**Also, an important note**, I am available this week but will be unavailable next week (Aug 13-20). I apologize for my absence but I really appreciate you all putting time into this project, it's turning out great!

Lawful2002 reviewed (approved):
Thanks, LGTM for codeowners files.

Lawful2002 commented on assets/i18n/en.json:
```diff
@@ -603,13 +603,19 @@
+    "I18N_FOOTER_NEWSLETTER_TEXT": "Be a part of the Oppia community! Our bi-weekly newsletter keeps you informed on new features, lessons and how Oppia is making a worldwide impact.",
     "I18N_FOOTER_OPPIA_FOUNDATION": "The Oppia Foundation",
+    "I18N_FOOTER_OPPIA_MISSION": "Oppia’s mission is to provide a learning platform that meets the unique needs for under resourced communities.",
+    "I18N_FOOTER_OPPIA_MISSION_DONATE_VOLUNTEER": "<a href=\"/donate\">Donate</a> or <a href=\"/volunteer\">Volunteer</a> today!",
```
Does this have to be an HTML snippet?

shricodev replied:
Maybe not necessarily, but since we're altering the color of the `Donate` and `Volunteer` anchor tags, not doing so will add some extra logic because we're using i18n instead of writing it in the HTML file itself.

Lawful2002 reviewed (commented):
Just one comment.

DubeySandeep commented:
Removing myself as @kevintab95 is already reviewing this PR!

shricodev commented:
> Hey All!
> 
> I apologize for the late reply on this. I have had some family issues and took the past few weeks off. **It looks like you've gotten a lot done, great job!**
> 
> The video of the breakpoints looks great. Also, @rfinch11 suggestion to use the button styling from the dropdown was a great idea. I think it looks better than the original design.
> 
> I did update the design file to tweak alignments. The newsletter signup, footer content and page content above it should all be the same width. Check it out and ask if you have any questions. Just make sure to tag me in any comments you need a direct response from.
> 
> **-Screenshot-** <img alt="Screenshot 2023-08-09 at 10 39 56 AM" width="612" src="https://user-images.githubusercontent.com/11729306/259465250-b141acdf-58d3-4774-9b78-f106b6126ad9.png">
> 
> **Also, an important note**, I am available this week but will be unavailable next week (Aug 13-20). I apologize for my absence but I really appreciate you all putting time into this project, it's turning out great!

Hey @ConorMack, I implemented your new Figma design. PTAL :)

Large - XL

![Screenshot 2023-08-11 113112](https://github.com/oppia/oppia/assets/76906722/983a9ffd-9d88-41b1-9a48-8bdc3bb172e0)

Medium - MD (I shortened the input since it may be unnecessarily wider, but can revert if you request a change.)

![Screenshot 2023-08-11 113216](https://github.com/oppia/oppia/assets/76906722/faaf9c39-47c3-4391-a7d8-a3b64347a2ad)


Small - S

![Screenshot 2023-08-11 113307](https://github.com/oppia/oppia/assets/76906722/43bdda50-2290-4117-9fba-f1e2ac2c98d7)

seanlip commented on core/platform/bulk_email/mailchimp_bulk_email_services.py:
```diff
@@ -252,18 +252,20 @@ def add_or_update_user_status(
+            'merge_fields': {}
         }
 
+        if merge_fields['NAME'] is not None:
```
Wouldn't this result in a KeyError if there is no NAME key in merge_fields?

shricodev replied:
Absolutely, Thank goodness you caught that! It was a rather silly mistake on my part. Thank you very much.
I have made the necessary change, PTAL

seanlip reviewed (changes requested):
Hi @shricodev, just one inline comment, thanks!

Also, could you please replace the screenshots/videos in the top-comment description with the updated ones? You don't need to keep the old ones, just show the latest UI there.

Thanks!

seanlip reviewed (commented):
Thanks @shricodev. No code comments, but in Conor's response, I think he said "The newsletter signup, footer content and page content above it should all be the same width." But in your video, at the start, the footer seems a lot wider than the page content? This should probably be fixed.

shricodev commented:
Hello, @seanlip, I have aligned the footer newsletter and footer content to the same width. However, I do have a concern. Will we be able to keep the footer width consistent with the width of the page content? This is because I've noticed varying page content widths across different Oppia pages. So for that reason, I have given a generic width for the component.

![Screenshot 2023-08-14 232515](https://github.com/oppia/oppia/assets/76906722/2186940f-9092-41a2-95ad-998506e08b0c)

![Screenshot 2023-08-14 232538](https://github.com/oppia/oppia/assets/76906722/7753c7be-1779-4a22-a66b-3cf8cf596146)
![Screenshot 2023-08-14 232613](https://github.com/oppia/oppia/assets/76906722/45ea93a0-5595-4f5f-bf8e-8e5c5b6875d9)

For example, In these three pages, we have different widths for the page content. Correct me if I am wrong. Thanks :)

seanlip commented:
@shricodev That's a good question. The Partnership page is incorrectly implemented and needs to be redesigned so that the content fits within a center portion of the screen even if the screen is wider. I suggest you take the /donate page as your basis, it seems roughly right in terms of having a clear max-width for the content.

kevintab95 reviewed (approved):
I was assigned per https://github.com/oppia/oppia/pull/18694#issuecomment-1671351797

However, it looks like there are no jobs or feature flags modified in this PR, so I'm happy to approve. Thanks!

shricodev commented:
> @shricodev That's a good question. The Partnership page is incorrectly implemented and needs to be redesigned so that the content fits within a center portion of the screen even if the screen is wider. I suggest you take the /donate page as your basis, it seems roughly right in terms of having a clear max-width for the content.

Hey @seanlip, Updated. PTAL

https://github.com/oppia/oppia/assets/76906722/6ca69e23-573a-4915-9bd5-ce9405fed981

PS: The test failure seems to be irrelevant to this PR. It seems to have some problems when installing dependencies.

seanlip reviewed (approved):
Thanks @shricodev, LGTM from my end! Just make sure to test a range of pages on your local  machine to ensure it's looking good everywhere. I'll restart the tests, but please contact the dev workflow team if there are still issues.

shricodev commented:
> Thanks @shricodev, LGTM from my end! Just make sure to test a range of pages on your local machine to ensure it's looking good everywhere. I'll restart the tests, but please contact the dev workflow team if there are still issues.

Hello @seanlip, While testing, I encountered an issue with the splash page. It does not seem to apply the CSS from `oppia-footer-component` correctly and instead applies the CSS from `oppia.css` I am unsure of the reason for this.

https://github.com/oppia/oppia/assets/76906722/fb2c8ad4-7690-481c-a851-86204616115b

One possible fix for that anchor tag styling to also take effect in the `splash-page` is doing it this way. Shall we go with this approach?

```html
    <p class="oppia-donate-volunteer">
      <a href="/donate">Donate</a> or <a href="/volunteer">Volunteer</a> today!
    </p>
```

instead of 

https://github.com/oppia/oppia/blob/b8852f9b8dd54eff616cfab9d5bec89cb08262fb/core/templates/base-components/oppia-footer.component.html#L31-L33


Also, when I switch the language to Arabic, it starts using the expected CSS. Is this intended or am I missing something? Could you please help me with the implementation?

Also, when subscribing to the newsletter from the splash page, it seems to not show the toastr message. I will probably fix this.

https://github.com/oppia/oppia/assets/76906722/e75a05b2-9cf3-4fe4-9c52-d4ea6acab29d


All the issues mentioned seem to happen only on the `splash-page` 

I think we should probably fix all these issues before merging the PR. Thanks :P

seanlip commented:
Hm, I'm not sure what the issue is @shricodev, but thanks for testing and checking this. It'll be good to fix this before the PR is merged.

I wonder if this is related to the lightweight-oppia-root vs oppia-root pages. If so, then @srijanreddy98 or @vojtechjelinek might be able to help / offer advice. But you might need to do a bit more digging into what's different between the splash and other pages, and write a [debugging doc](https://github.com/oppia/oppia/wiki/Debugging-Docs).

vojtechjelinek commented:
> Hm, I'm not sure what the issue is @shricodev, but thanks for testing and checking this. It'll be good to fix this before the PR is merged.
> 
> I wonder if this is related to the lightweight-oppia-root vs oppia-root pages. If so, then @srijanreddy98 or @vojtechjelinek might be able to help / offer advice. But you might need to do a bit more digging into what's different between the splash and other pages, and write a [debugging doc](https://github.com/oppia/oppia/wiki/Debugging-Docs).

I can help with the bug but please prepare a debugging doc.

shricodev commented:
@vojtechjelinek I have mentioned almost everything the same in the debugging doc. PTAL [
debugging doc link](https://docs.google.com/document/d/1FI08NaAT9HXX-i-fjrqlIj0Q9xGy86W7XUmcgzeCoaI/edit?usp=sharing)

I would highly appreciate your help on this. Thanks

vojtechjelinek commented:
@shricodev I cannot comment on your doc. Two initial things I think you should try to do:
 1. Investigate why the CSS file for the footer is not loaded.
 2. Investigate what line of code that you changed lead to the problem, like for example if you check the splash page without your changes does it load the CSS?

shricodev commented:
@vojtechjelinek  I have updated the permissions for the doc. The problem I have noticed is that the footer anchor tag styles are overridden by external CSS file `oppia.css` or not applied to this page. I have tried all the workarounds but I am not sure why. I don't think any change in my code has actually led to this issue.

Would you mind fixing it, please? :p

shricodev commented:
@vojtechjelinek  I have updated the permissions for the doc. The problem I have noticed is that the footer anchor tag styles are overridden by the external file `oppia.css` or not applied to this page. I have tried all the workarounds but I am not sure why. I don't think any change in my code has actually led to this issue.

Would you fix this issue, please :p

vojtechjelinek commented:
> @vojtechjelinek I have updated the permissions for the doc. The problem I have noticed is that the footer anchor tag styles are overridden by external CSS file `oppia.css` or not applied to this page. I have tried all the workarounds but I am not sure why. I don't think any change in my code has actually led to this issue.

Wait so the same error also happens on develop branch? Also document in the debugging doc wat workarounds have you actually tried that is the only way I can help.

> Would you mind fixing it, please? :p

Sorry I do not have the time to fix it myself, you need to try the possible ways, document them and I can provide help with what you can try next.

shricodev commented:
> > @vojtechjelinek I have updated the permissions for the doc. The problem I have noticed is that the footer anchor tag styles are overridden by external CSS file `oppia.css` or not applied to this page. I have tried all the workarounds but I am not sure why. I don't think any change in my code has actually led to this issue.
> 
> Wait so the same error also happens on develop branch? Also document in the debugging doc wat workarounds have you actually tried that is the only way I can help.
> 
> > Would you mind fixing it, please? :p
> 
> Sorry I do not have the time to fix it myself, you need to try the possible ways, document them and I can provide help with what you can try next.

I fixed the issue by breaking down that long snippet into multiple tags. The problem was that the splash page was not recognizing the code nested in the i18n as a proper tag.

seanlip commented:
@shricodev I'm not sure your solution works, wouldn't it result in "or" and "today!" not being translated?

In general it's considered not good practice to split sentences/phrases for translation, see https://github.com/oppia/oppia/wiki/How-to-develop-for-i18n#good-i18n-practices

shricodev commented:
> 

@seanlip so, shall we add separate i18n for these?

seanlip commented:
Generally there should be a single i18n string for the overall sentence. It's fine to have HTML in it, try to keep the HTML as simple as possible. (E.g. define CSS using an outer container's class instead.) See the link I gave in the previous comment for best practices.

If you need further guidance, a debugging doc is needed so that we can discuss this efficiently.

shricodev commented:
> Generally there should be a single i18n string for the overall sentence. It's fine to have HTML in it, try to keep the HTML as simple as possible. (E.g. define CSS using an outer container's class instead.) See the link I gave in the previous comment for best practices.
> 
> If you need further guidance, a debugging doc is needed so that we can discuss this efficiently.

The same was my approach before https://github.com/oppia/oppia/pull/18694/commits/bff20162bb82587cec8bfa0f11c60ae1aa3cd153 commit, but with that approach, styles were being overridden. I have updated the debugging doc. PTAL

https://docs.google.com/document/d/1FI08NaAT9HXX-i-fjrqlIj0Q9xGy86W7XUmcgzeCoaI/edit?usp=sharing

seanlip commented:
Thanks, I left a comment on your doc. Let's continue the discussion there until it's resolved.

shricodev commented:
The test failure seems to be irrelevant to the PR changes!

srijanreddy98 commented on core/templates/base-components/oppia-footer.component.ts:
```diff
@@ -16,22 +16,31 @@
   selector: 'oppia-footer',
   templateUrl: './oppia-footer.component.html',
-  styleUrls: ['./oppia-footer.component.css']
+  styleUrls: ['./oppia-footer.component.css'],
+  encapsulation: ViewEncapsulation.None,
```
ViewEncapsulation set to None is not something we like to go for. This can lead to a whole host of issues if not done properly. For a component that uses ViewEncapsulation as None, the css styles need to be manually scoped. It is usually achieved by wrapping the entire component in a div with a class named after the component and using that class to scope the css. However, doing that on a top level component is too risky as we may end up affecting lower level non encapsulated components. Please try to move the problematic code (I assume links with translation were causing the problem) to a smaller re-usable component with no view encapsulation and the correct manually scoped.

shricodev replied:
I have moved those translations to a separate reusable component. PTAL

srijanreddy98 commented on core/templates/base-components/footer-donate-volunteer.component.html:
```diff
@@ -0,0 +1,25 @@
+      line-height: 24px;
+    }
+  }
+</style>
```
Looks like you are missing a new line here. Can you also get the RTL version checked once with you latest changes?

shricodev replied:
PTAL
![rtl](https://github.com/oppia/oppia/assets/76906722/9339007e-16e1-4003-8d8d-94aacabc095f)

shricodev commented:
@seanlip PTAL

seanlip commented:
@shricodev Could you please address the CI issues and then reassign this to me (if you still need review) or merge directly (if you do not)? Thanks.

shricodev commented:
The frontend tests seem to fail when installing the dependencies, especially https://nodejs.org/dist/v16.13.0/node-v16.13.0-linux-x64.tar.gz.

The e2e coverage and TypeScript check seem to fail for some other change (610 errors found during compilation), which is definitely not related to this PR. Could you please take a look? Maybe these issues were introduced in some recently merged PRs.

Ref: https://github.com/oppia/oppia/actions/runs/6573269549/job/17855979111?pr=18694

seanlip commented:
I can't take a look, I'm travelling. Does this happen on develop and on
other PRs?

On Thu, Oct 19, 2023, 1:12 PM Shrijal Acharya ***@***.***>
wrote:

> Assigned #18694 <https://github.com/oppia/oppia/pull/18694> to @seanlip
> <https://github.com/seanlip>.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia/pull/18694#event-10710623999>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACQV5SRZ5MEBALLFKQFBUCLYAEKMXAVCNFSM6AAAAAA2ZUWQB2VHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJQG4YTANRSGM4TSOI>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>
